cmd/clusterctl/pkg/client/repository/template.go

```diff
@@ -45,9 +47,85 @@ type Template interface {
 	Variables() []string
 
 	// TargetNamespace where the template objects will be installed.
-	// This value is inherited by the template options.
 	TargetNamespace() string
 
-	// Yaml file defining all the cluster objects.
-	Yaml() []byte
+	// Yaml file defining all the cluster template objects.
+	Yaml() ([]byte, error)
+
+	// Objs return the cluster template  in the form of a list of Unstructured objects.
+	Objs() []unstructured.Unstructured
+}
+
+// template implement Template
+type template struct {
+	config.Provider
+	version         string
+	flavor          string
+	bootstrap       string
+	variables       []string
+	targetNamespace string
+	objs            []unstructured.Unstructured
+}
+
+// Ensures template implement the Template interface.
+var _ Template = &template{}
+
+func (t *template) Version() string {
+	return t.version
+}
+
+func (t *template) Flavor() string {
+	return t.flavor
+}
+
+func (t *template) Bootstrap() string {
+	return t.bootstrap
+}
+
+func (t *template) Variables() []string {
+	return t.variables
+}
+
+func (t *template) TargetNamespace() string {
+	return t.targetNamespace
+}
+
+func (t *template) Objs() []unstructured.Unstructured {
+	return t.objs
+}
+
+func (t *template) Yaml() ([]byte, error) {
+	return util.FromUnstructured(t.objs)
+}
+
+// newTemplate returns a new objects embedding a cluster template YAML file.
+func newTemplate(provider config.Provider, version, flavor, bootstrap string, rawyaml []byte, configVariablesClient config.VariablesClient, targetNamespace string) (*template, error) {
+	// Inspect variables and replace with values from the configuration.
+	variables := inspectVariables(rawyaml)
+
+	yaml, err := replaceVariables(rawyaml, variables, configVariablesClient)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to perform variable substitution")
+	}
+
+	// Transform the yaml in a list of objects, so following transformation can work on typed objects (instead of working on a string/slice of bytes).
+	objs, err := util.ToUnstructured(yaml)
```

Comment:
Are we documenting what is appropriate to put in a cluster template? I assume all k8s resources, capi resources are supported. What happens when someone adds an arbitrary resource (say they have some custom CRD)?

Reply:
> Are we documenting what is appropriate to put in a cluster template?

There is https://github.com/kubernetes-sigs/cluster-api/pull/1994 that documents what assumptions clusterctl makes on the template file.

quoting: _"The cluster templates file, are single YAML with _all_ the objects required to create a new workload cluster. "_

> What happens when someone adds an arbitrary resource (say they have some custom CRD)?

clusterctl assumes all the required CRDs should be in the components YAML that is installed when the provider is installed.

Quoting: _"(the components YAML) is a single YAML with _all_ the components required 
for installing the provider itself (CRD, Controller, RBAC etc.)"_

Nothing prevents you to do in a different way, but if you do so, make sure that everything is in place when required